packages/components/src/utils/hooks/use-focus-outside.js

```diff
@@ -0,0 +1,211 @@
+	useEffect( () => {
+		if ( ! onFocusOutside && ! __unstableNodeRef?.current ) {
+			cancelBlurCheck();
+		}
+	}, [ onFocusOutside, __unstableNodeRef ] );
+
+	/**
+	 * Handles a mousedown or mouseup event to respectively assign and
+	 * unassign a flag for preventing blur check on button elements. Some
+	 * browsers, namely Firefox and Safari, do not emit a focus event on
+	 * button elements when clicked, while others do. The logic here
+	 * intends to normalize this as treating click on buttons as focus.
+	 *
+	 * @see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus
+	 *
+	 * @param {SyntheticEvent} event Event for mousedown or mouseup.
+	 */
+	const normalizeButtonFocus = useCallback(
+		( event ) => {
+			const { type, target } = event;
+			const isInteractionEnd = includes(
+				[ 'mouseup', 'touchend' ],
+				type
+			);
+
+			if ( isInteractionEnd ) {
+				preventBlurCheck.current = false;
+			} else if ( isFocusNormalizedButton( target ) ) {
+				preventBlurCheck.current = true;
+			}
+		},
+		[ preventBlurCheck ]
+	);
+
+	/**
+	 * A callback triggered when a blur event occurs on the element the handler
+	 * is bound to.
+	 *
+	 * Calls the `onFocusOutside` callback in an immediate timeout if focus has
+	 * move outside the bound element and is still within the document.
+	 *
+	 * @param {SyntheticEvent} event Blur event.
+	 */
+	const queueBlurCheck = useCallback(
+		( event ) => {
+			// React does not allow using an event reference asynchronously
+			// due to recycling behavior, except when explicitly persisted.
+			event.persist();
+
+			// Skip blur check if clicking button. See `normalizeButtonFocus`.
+			if ( preventBlurCheck.current ) {
+				return;
+			}
+
+			blurCheckTimeoutId.current = setTimeout( () => {
+				// If document is not focused then focus should remain
+				// inside the wrapped component and therefore we cancel
+				// this blur event thereby leaving focus in place.
+				// https://developer.mozilla.org/en-US/docs/Web/API/Document/hasFocus.
+				if ( ! document.hasFocus() ) {
+					event.preventDefault();
+					return;
+				}
+
+				if ( 'function' === typeof onFocusOutside ) {
+					onFocusOutside( event );
+				} else if (
+					'function' ===
+					typeof __unstableNodeRef?.current?.handleFocusOutside
+				) {
+					// Call the legacy `handleFocusOutside` method if defined
+					// as a fallback for the `withFocusOutside` HOC.
+					const node = __unstableNodeRef.current;
+					const callback =
+						__unstableNodeRef.current.handleFocusOutside;
+					callback.call( node, event );
+				}
+			}, 0 );
+		},
+		[
+			preventBlurCheck,
+			blurCheckTimeoutId,
+			onFocusOutside,
+			__unstableNodeRef,
+		]
+	);
+
+	return {
+		onFocus: cancelBlurCheck,
+		onMouseDown: normalizeButtonFocus,
+		onMouseUp: normalizeButtonFocus,
+		onTouchStart: normalizeButtonFocus,
+		onTouchEnd: normalizeButtonFocus,
+		onBlur: queueBlurCheck,
```

Comment:
A potential issue would be if the element receiving these props also has some of them defined, then these will override them. We can expect the developer to group the event handlers by themselves, or we can do it for them here, I'm okay with both approaches.

Reply:
Yeah, I don't like making the implementer handle so many callbacks, but I don't think there's a good alternative. Using DOM event handlers and a ref would be possible if the hook didn't have to handle virtual event bubbling, but it does.

What are the options for improving this? I've seen the pattern where the hook takes and returns the props from a component, which is slightly easier to use:
```jsx
const propsWithFocusOutside = useFocusOutside( props, onFocusOutside );

return (
    <div { ...propsWithFocusOutside }>
```

Is that what you mean when you say we can do it for them?

Reply:
Wait, nvm. I was thinking component design 😅 .

I think this is fine and I don't think we should pass `props` to `useFocusOutside`. However, I think we can try to refactor this to a component instead.

```js
<FocusOutside as="div">
  // ...
</FocusOutside>
```

And we can then handle the composition/grouping of the event handlers for them.

```js
<FocusOutside as="div" onMouseDown={() => console.log("mousedown")}>
  // ...
</FocusOutside>

function FocusOutside(props) {
  // ...
  return {
    // ...
    onMouseDown: groupEvents(normalizeButtonFocus, props.onMouseDown)
  };
}
```

Reply:
Hah 😄 

Yeah, it's an interesting idea making it a component. I definitely need to think about future usage here. My original goal was to remove the wrapping `div` introduced in the combobox component in https://github.com/WordPress/gutenberg/pull/27367.

But within Popovers or Modals this will likely be used with some other HOCs/components:
- `withFocusReturn` (another HOC that renders a `div`)
- `withConstrainedTabbing` (yet again another HOC that renders a `div`)
- `IsolatedEventContainer` (a wrapper that renders a `div`)

That's why these components end up with so many `div` wrappers when they render, and it'd be nice to reduce that as much as possible with a consistent approach (refactor to hooks/components).

Does the `as` pattern scale well to multiple components that render the same element? I think that's where hooks might be better suited to the problem and more explicit

Reply:
Oh this is interesting, thx for the additional info!

I think in theory we can still do it with `as` prop:

```js
<FocusOutside as={FocusReturn}>
  // ...
</FocusOutside>
```

But this can easily become tedious if we want to combine from multiple sources. Hooks with a `useCompositeProps` hook could solve this.

```js
const htmlProps = useCompositeProps(
  useFocusOutside(),
  useFocusReturn(),
  useConstrainedTabbing(),
  // ...
);

<div {...htmlProps></div>
```

But this could also be a little bit overly complicated.

Maybe just export `mergeEventHandlers` to the user so that they can combine them by themselves?

Might be worth studying how [reakit](https://reakit.io/docs/composition/#props-hooks) solves this though.